README.md
<h1 align="center">
  Teemo(J)S
</h1>
<p align="center">
  <em>Now in TypeScript</em>
</p>
<p align="center">
  <a href="https://github.com/MingweiSamuel/TeemoJS/"><img src="https://cdn.communitydragon.org/latest/champion/Teemo/square" width="20" height="20" alt="Github"></a>
  <a href="https://www.npmjs.com/package/teemojs"><img src="https://img.shields.io/npm/v/teemojs.svg?style=flat-square&logo=npm" alt="NPM"></a>
  <a href="https://packagephobia.now.sh/result?p=teemojs%40next"><img src="https://flat.badgen.net/packagephobia/install/teemojs@next" alt="Install Size"></a>
  <a href="https://bundlephobia.com/result?p=teemojs%40next"><img src="https://flat.badgen.net/bundlephobia/min/teemojs%40next" alt="Minified Size"></a>
</p>

Teemo(J)S is a fast, lightweight, and fully-typed Riot API wrapper written in
TypeScript,
<!-- with additional support for DDragon, CDragon, and
[meraki-analytics/kernel](https://github.com/meraki-analytics/kernel). -->
Contained in about 500 lines, it has a minimalist design to make it flexible
and easy to maintain.

Features:

- Fast, efficient, correct rate limiting.
- Automatic retries.
- TypeScript typings:
  - Argument & region type checking on API calls.
  - Fully-typed response DTOs, supporting `strictNullChecks`.
- Up-to-date (or if not, [make an issue!](https://github.com/MingweiSamuel/TeemoJS/issues/new?title=Update+needed&body=%3CSOMETHING%3E+is+missing.)).

## Installation

```sh
npm install --save teemojs
```

## Usage

### Example
```javascript
const { TeemoApi, PlatformRoute, RegionalRoute } = require('teemojs');
const api = TeemoApi.createRiotApi('RGAPI-KEY-HERE').proxy();

async function main() {
  const summoner = await api.summonerV4.getBySummonerName(PlatformRoute.NA1, {
    path: { summonerName: 'x blotter' },
  });
  console.log(`${summoner.name}'s account id is ${summoner.accountId}.`);

  // Get summoner's games on Teemo and Illaoi for a particular season.
  const matchlist = await api.matchV5.getMatchlist(RegionalRoute.AMERICAS, {
    path: [ summoner.puuid ],
    query: { queue: 420 },
  });
  console.log(`Fetched ${matchlist.matches.length} games.`);

  // ...
}

main();
```

### Syntax

There are two equivalent syntaxes:
- `req` syntax, all API calls are done via the `.req(endpoint, method, ...)` method.
- `proxy` syntax, endpoints and methods can be called as fields: `api.endpoint.method(...)`.

The later uses [`Proxy`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy)
to call the former.


**`req` syntax:**
```javascript
const api = TeemoApi.createRiotApi('RGAPI-KEY-HERE');
aPromise = api.req(endpoint, method, region, {
  path: pathParams,
  query: queryParams,
  body: bodyParam,
});
```
**`proxy` syntax:**
```javascript
const api = TeemoApi.createRiotApi('RGAPI-KEY-HERE')
    .proxy(); // Convert to proxy.
aPromise = api.endpoint.method(region, {
  path: pathParams,
  query: queryParams,
  body: bodyParam,
});
```

#### Parameters
- `endpoint` - An API endpoint in lower camel case, such as `summonerV4` or
  `matchV5`.
- `method` - An method in `endpoint`, such as `getBySummonerName` or
  `getMatchlist` (respectively).
  - See [`configs.ts`](https://github.com/MingweiSamuel/TeemoJS/blob/master/src/configs.ts)
  for a full list of endpoints and methods.
- `region` - The region to request to, from the `Region` enum (`Region.NA1`) or
  a string.
- The parameters object. This can be omitted if no parameters are needed, e.g.
  for `lolStatusV3` `getShardData`.
  - `path` - If needed, an array or object dictionary of parameters. For the
    dictionary, key names must match the `{pathParam}` names (no curly braces).
  - `query` - If needed, an object dictionary of query parameter names mapping
    to values or arrays of values.
  - `body` - If needed, a JSON value, to be serialized.

#### Return Value & Error Handling
Returns `dataPromise`, the data returned by the Riot API. This is the
`JSON.parse`d body of the response. This is:
- A `Promise` resolving to the JSON value parsed from the response body on
  success.
- A `Promise` resolving to `null` if the response had no body (status code 204,
  404, 422).
- A `Promise` rejecting with an `Error`. If a response was received, the
  `Error` will have a `.response` field containing the failed repsonse.

Throws an `Error` immediately (synchronously) if the request could not be made.
This happens if the `endpoint`, `method` pair were invalid, or path parameters
were missing.

### In-Browser
After `npm install`ing, `node_modules/teemojs/dist/browser.js` contains a
browser-ready version of TeemoJS. However, the Riot Games API does not support
cross-origin requests (CORS), so you will need to configure a proxy such as
[meraki-analytics/kernel](https://github.com/meraki-analytics/kernel). There is
currently no out-of-the-box config for this, but it is planned.

<!--Usable with `TeemoJS.ddragonConfig`, `TeemoJS.cdragonConfig`, and `TeemoJS.kernelConfig`
([meraki-analytics/kernel](https://github.com/meraki-analytics/kernel)). (`defaultConfig` will not work due to CORS).

This is mainly intended for development, and it uses ES6 features.
For production, you should include `src/index.js` and your needed config(s)
(`config/<config>.json`) in your webpack or other bundle.-->

### Other features

#### Riot API with separate keys for TFT/LOR/Tournament APIs

```javascript
const api = TeemoApi.createRiotApi({
  default: 'RGAPI-LOL-KEY-HERE',
  tft: 'RGAPI-OPTIONAL-TFT-KEY-HERE',
  lor: 'RGAPI-OPTIONAL-LOR-KEY-HERE',
  tournament: 'RGAPI-OPTIONAL-TOURNAMENT-KEY-HERE',
});
// Use as normal.
```

#### TypeScript typings

```typescript
import { TeemoApi, PlatformRoute, RegionalRoute, summonerV4, matchV5 } from "TeemoJS";
// ...
const summoner: summonerV4.SummonerDTO = await api.summonerV4.getBySummonerName(PlatformRoute.NA1, {
  path: { summonerName: 'x blotter' },
});
const matchlist: matchV5.MatchlistDto = await api.matchV5.getMatchlist(RegionalRoute.AMERICAS, {
  path: [ summoner.puuid ],
  query: { queue: 420 },
});
```
```typescript
let myResult: ReturnType<typeof api.championMasteryV4.getAllChampionMasteries>;
// myResult: Promise<championMasteryV4.ChampionMasteryDTO[]>
```


<!--
#### Riot API with separate key for TFT endpoints
```javascript
// Makes a deep copy to not modify the original (optional).
const config = JSON.parse(JSON.stringify(TeemoJS.defaultConfig));
config.key = riotApiKey;
config.endpoints.tft['*'] = { key: tftApiKey }; // Use tftApiKey for tft endpoints.
const api = TeemoJS(config);
// Use as normal.
```

#### Data Dragon: `TeemoJS.ddragonConfig`
```javascript
const api = TeemoJS(TeemoJS.ddragonConfig);
const data = await api.req('cdn.championByKey', [ version, 'en_US', 'Teemo' ]);
```

#### Community Dragon: `TeemoJS.cdragonConfig`
```javascript
const api = TeemoJS(TeemoJS.cdragonConfig);
const data = await api.req('cdn.champion', { patch: 'latest', champion: 'monkeyking' });
```

#### meraki-analytics/kernel: `TeemoJS.kernelConfig`
Local development on port 8080:
```javascript
// Make a deep copy to not modify the original (optional).
const config = JSON.parse(JSON.stringify(TeemoJS.kernelConfig));
config.origin = "http://localhost:8080"
const api = TeemoJS(config);
// Use as normal.
```


## Configuration

The `TeemoJS` constructor can take an second argument which is a configuration object.
You should use `TeemoJS.defaultConfig` (or a different available configuration)
as a starting point and override any fields as neccesary.

### Available configurations
- [`TeemoJS.defaultConfig`](https://github.com/MingweiSamuel/TeemoJS/blob/master/config/default.json):  
  Default Riot Games API config.
- [`TeemoJS.emptyConfig`](https://github.com/MingweiSamuel/TeemoJS/blob/master/config/empty.json):  
  Empty (no endpoints) Riot Games API config. Probably not that useful.
- [`TeemoJS.ddragonConfig`](https://github.com/MingweiSamuel/TeemoJS/blob/master/config/ddragon.json):  
  Data Dragon config ([example](http://ddragon.leagueoflegends.com/cdn/9.7.1/data/cs_CZ/champion/Amumu.json)).
- [`TeemoJS.cdragonConfig`](https://github.com/MingweiSamuel/TeemoJS/blob/master/config/cdragon.json):  
  Community Dragon config ([example](https://raw.communitydragon.org/latest/plugins/rcp-be-lol-game-data/global/default/v1/
- [`TeemoJS.kernelConfig`](https://github.com/MingweiSamuel/TeemoJS/blob/master/config/kernel.json):  
  [meraki-analytics/kernel](https://github.com/meraki-analytics/kernel) config. Allows you to use `kernel` as an API
  proxy while still having TeemoJS's familiar interface in your browser webpage. Must set `origin` to wherever
  kernel is running on.
- ~~TeemoJS.championGGConfig~~ (RIP).
- (More in the future?)

### Example constructors with `config`
```javascript
let api = TeemoJS('RGAPI-KEY-HERE', { ...TeemoJS.defaultConfig, maxConcurrent: 100 });
let api = TeemoJS(configWithKey);
```

### `config` Object

- `retries` [int]: Number of times to retry request if the request fails with a retriable error. Zero for no retires.
- `maxConcurrent` [int]: Maximum live requests to allow simultaneously.
- `distFator` [float 0..1]: Factor to multiply rate limits by. This can be changed at any time using the `api.setDistFactor(x)` method. For example, if your API key was distributed across two computers, you could set this to 0.5.
- `key` [OPTIONAL string]: Overrides the `key` argument passed into the constructor. Do not set unless you actually use this key.

#### Rate limit configuration

- `bucketsConfig` [object]: Arguments passed into `TokenBucket` constructor when rate limits are first detected (or
  change). May have any subset of the following fields:
  - `bins` [OPTIONAL integer, default 20]: Number of bins to compute circular buffer with.
  - `binFactor` [OPTIONAL number in `(0, 1]`, default 0.95): Fraction of requests that can go in each bin. 0.95 means 95% of
    requests can be bursted in one bin (the remaining 5% can be used in the next bin). Must be in range.
  - `overhead` [OPTIONAL number (milliseconds), default 20ms): Time in milliseconds to expand the bucket by. Helps deal
    with latency causing requests to be counted in the wrong bucket. Increase this if you are getting 429s when time
    changes from one bucket to the next.

#### "Internal" configurations

The following properties are mainly used for _dealing_ with all the different behaviors of the different APIs.
A lot of their exact behaviors are "self-documented" in the code :).

- `keyPath` \[OPTIONAL string\]: JSON path the key will be inserted into in the `reqConfig`. Or `null` if no key needed.
- `origin` \[interpolated string\]: String containing the protocol and host without a trailing forward slash. May have `{}`
  `{name}`s for `originParams`.
- `regionPath` \[OPTIONAL string\]: JSON path the region will be inserted into in the `reqConfig`. Or `null` if no region needed.
- `defaultBuckets` \[object\[\]\]: Array of configuration options for a rate limit's default buckets. These buckets remain in use until the actual rate limit is detected via headers. Each must have at least `timespan` (in milliseconds) and `limit`.
- `rateLimitTypeApplication` \[object\]: Rate limit type object for the application rate limit.
  Containing strings `name`, `headerLimit`, and `headerCount`.
  `name` is the name of the rate limit used for detecting which type caused a 429.
  `headerLimit` and `headerCount` are header names for the max rate limit and rate limit count respectively.
- `rateLimitTypeMethod` \[object/null\]: Same as `rateLimitTypeApplication` but for method rate limits. May be null if the API does not have per-method rate limits.
- `defaultRetryAfter` [string/null]: Default retry after in seconds if the `headerRetryAfter` is not provided in a 429. Use `null` to cause missing `headerRetryAfter` headers to throw an error.
- `headerRetryAfter` [string/null]: Header name to look for retry after time in seconds when a 429 is hit.
  If `null`, `defaultRetryAfter` should be set.
- `headerLimitType` [string/null]: Header name to match with the `name` of a rateLimitType to determine which rate
  limit hit a 429. `null` means always use `defaultLimitType`.
- `defaultLimitType` [string/null]: Default `name` value to use when the API doesn't return which limit is hit.
  Set to `null` to throw an error if a 429 happens for no reason. Set to `rateLimitTypeXYZ.name` to default to rate limit type XYZ.
- `collapseQueryArrays` [boolean]: If `false`, query arrays will be represented as `a=1&a=2&a=3`.
  If `true`, `a=1,2,3` will be used. Riot API uses the former, champion.gg used the later (but is dead).
- `endpoints` [nested object]: A nested object structure where the leaf values are `reqConfig`s.
  Objects may be nested to any level for organizational purposes.
  When using the API, the period-delimited path is supplied.
  Can use `'*'` wildcards for config to apply to multiple endpoints.
  - `path` \[interpolated string\]: URL path with leading slash. Technically optional but that would be silly.
  - `fetch`: [Fetch API `init` config](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch#Parameters),
    second argument passed into `fetch` function.
  - `key` \[OPTIONAL string\]: Per-endpoint(s) API key. Useful for using separate key for certain (*cough* TFT) endpoints.
  - `origin` \[OPTIONAL string\]: URL origin (protocol and hostname).
  - `originParams` \[OPTIONAL object or array\]: for interpolating `origin`.
  - `pathParams` \[OPTIONAL object or array\]: should probably be unset, as will be partially overriden by `req` arguments.
  - `queryParams` \[OPTIONAL object\].
  - (Other fields?)
-->


CHANGELOG.md
# 3.0.0 (Prerelease)

## Non-API changes
* Teemo(J)S is now written in TypeScript and is fully-typed.
* Teemo(J)S is now `MIT`-licensed (switched from `LGPL-3.0`).

<!--
<table>
<tr><th>Before (1.x.x)</th><th>After (3.x.x)</th></tr>
<tr></tr>
<tr><td><pre lang="javascript">
summonerFuture = api.get(
  'na1', 'summoner.getBySummonerName',
  'xBlotter');
</pre></td>
<td><pre lang="javascript">
summonerFuture = api.req(
  'na', 'lol.summonerV4.getBySummonerName', 'xBlotter');
</pre></td></tr>
<tr></tr>
<tr><td><pre lang="javascript">
future = api.get(
  'euw1', 'league.getLeagueEntries',
  'RANKED_SOLO_5x5', 'GOLD', 'IV',
  { page: 12 });
</pre></td>
<td><pre lang="javascript">
future = api.req(
  'euw', 'lol.leagueV4.getLeagueEntries',
  [ 'RANKED_SOLO_5x5', 'GOLD', 'IV' ],
  { page: 12 });
</pre>Or, using an object for path params:
<pre lang="javascript">
future = api.req(
  'euw', 'lol.leagueV4.getLeagueEntries',
  { queue: 'RANKED_SOLO_5x5',
    tier: 'GOLD', division: 'IV' },
  { page: 12 });
</pre></td></tr>
</table>
-->

# 2.0.0 (prerelease)

2.0 was never released and skipped for version 3.0.

Changes made for TeemoJS 2.0.0

<table>
<tr><th>Before (1.3.4)</th><th>After (2.0.0)</th></tr>
<tr></tr>
<tr><td><pre lang="javascript">
summonerFuture = api.get(
  'na1', 'summoner.getBySummonerName',
  'xBlotter');
</pre></td>
<td><pre lang="javascript">
summonerFuture = api.req(
  'na', 'lol.summonerV4.getBySummonerName', 'xBlotter');
</pre></td></tr>
<tr></tr>
<tr><td><pre lang="javascript">
future = api.get(
  'euw1', 'league.getLeagueEntries',
  'RANKED_SOLO_5x5', 'GOLD', 'IV',
  { page: 12 });
</pre></td>
<td><pre lang="javascript">
future = api.req(
  'euw', 'lol.leagueV4.getLeagueEntries',
  [ 'RANKED_SOLO_5x5', 'GOLD', 'IV' ],
  { page: 12 });
</pre>Or, using an object for path params:
<pre lang="javascript">
future = api.req(
  'euw', 'lol.leagueV4.getLeagueEntries',
  { queue: 'RANKED_SOLO_5x5',
    tier: 'GOLD', division: 'IV' },
  { page: 12 });
</pre></td></tr>
</table>

<!--
## Main Changes
* `get` renamed to `req` (as `POST`, `PUT`, etc. endpoints now exist).
  * First param is a region (not a platform), automatically converted per `lol.*`, `tft.matchV1.*`, `tournament.*`.
    * Regions are: `"br", "eune", "euw", "jp", "kr", "lan", "las", "na", "oce", "tr", "ru", "pbe", "americas", "europe", "asia"`.
  * Second param (`endpointPath`) now includes `lol`, `tft`, or `tournament` as an additional prefix as well as the
    version (`V4`).
    * `summoner.getBySummonerName` (before) -> `lol.summonerV4.getBySummonerName` (after)
    * `tftLeague.getLeagueById` (before) -> `tft.leagueV1.getLeagueById` (after)
    * (no support for tournament before) -> `tournament.v4.updateCode` (after)
  * Path param(s) are now a single param, which may be a single value, list of values, or object map of values.
    * Before: `..., 'RANKED_SOLO_5x5', 'GOLD', 'IV', ...`
    * After (array):  `..., [ 'RANKED_SOLO_5x5', 'GOLD', 'IV' ], ...`
    * After (object): `..., { queue: 'RANKED_SOLO_5x5', tier: 'GOLD', division: 'IV' }, ...`
  * Query argument is unchanged (object map).
  * Additional body argument can be provided for `POST`, `PUT`, etc. requests. Value will be `JSON.stringify`d.
* `TeemoJS` is now browser-ready.
  * After `npm install`ing, `node_modules/teemojs/dist/browser.js` is a browserified version.
  * Usable with `ddragonConfig`, `cdragonConfig`, and `kernelConfig`
    ([meraki-analytics/kernel](https://github.com/meraki-analytics/kernel)). (`defaultConfig` will not work due to CORS).

### New Configs
* `TeemoJS.ddragonConfig` - Config for `https://ddragon.leagueoflegends.com` (drop-in).
* `TeemoJS.cdragonConfig` - Config for `https://[raw|cdn].communitydragon.org` (drop-in).
* `TeemoJS.kernelConfig` - Config for [meraki-analytics/kernel](https://github.com/meraki-analytics/kernel) (user must
  set `origin` field).

## Other API Changes
* TeemoJS is now `MIT`-licensed (switched from `LGPL-3.0`).
* The first argument to `req` is now a region rather than a platform. This will be automatically transformed to the
  right platform for different endpoints. For example, you should use `'na'` instead of `'na1'`. In `tftMatchV1`, it
  will automatically be translated into `'americas'`.
* Endpoint paths now include the version number (`V1`, `V3`, `V4`, etc).
  For example, `api.get('na1', 'match.getMatch', 3022273777)` is now `api.req('na', 'lol.matchV4.getMatch', 3022273777)`.
  (Note that `matchV4` has been replaced by `matchV5` with different arguments/routes).
* On failed requests (when retries run out or the request is not retryable),
  an `Error` will always be thrown. Previously, some error codes (`403`, `401`, ?) would not result in `Error`s thrown.
* Thrown `Error`s may now contain a `.response` field containing the latest failed Fetch API `Response` instance.

### Config Changes
* `"prefix"` renamed to `"origin"`.
* `"bucketsConfig"` added for fine tuning `TokenBucket`s.
  * `bins` (default: 20): Number of bins to compute circular buffer with.
  * `binFactor` (default: 0.95): Fraction of requests that can go in each bin. 0.95 means 95% of requests can be bursted
    in one bin (the remaining 5% can be used in the next bin).
  * `overhead` (default: 20ms): Time in milliseconds to expand the bucket by. Helps deal with latency causing requests
    to be counted in the wrong bucket.
* `"keyPath"` added to determine where to put key. `keyHeader` and `keyQueryParam` have been removed.
* `"regionPath"` added to determine where to put region string.
* Formatted strings now use `{}` or `{nameHere}` instead of `%s`.
  (Python `str.format`-style instead of sprintf-style `%s`).
* `endpoints` are now more than just a path. Now fetch params,
  query params, and other things can be defined in addition to `path`.
* `endpoints` can now contain `'*'` paths which will match all child endpoints. API `key`, `regionTable`, `pathParams`,

index.js
'use strict';

const util = require("util");
const req = require("request-promise-native");
([ 'emptyConfig', 'defaultConfig', 'championGGConfig' ].forEach(
  config => RiotApi[config] = require('./' + config + '.json')));
const DEBUG = typeof global.it === 'function';
const delayPromise = delay => new Promise(resolve => setTimeout(resolve, delay));

/** `RiotApi(key [, config])` or `RiotApi(config)` with `config.key` set. */
function RiotApi(key, config = {}) {
  if (!(this instanceof RiotApi)) return new RiotApi(...arguments);
  this.config = JSON.parse(JSON.stringify(RiotApi.defaultConfig));
  if (key instanceof Object)
    config = key;
  else
    this.config.key = key;
  Object.assign(this.config, config);
  this.regions = {};
  this.hasRegions = this.config.prefix.includes('%s');
}
RiotApi.prototype.get = function() {
  if (!this.hasRegions)
    return this._getRegion(null).get(this.config.prefix, ...arguments);
  let [ region, ...rest ] = arguments;
  rest.unshift(util.format(this.config.prefix, region));
  return this._getRegion(region).get(...rest);
};
/**
 * Limits requests to FACTOR fraction of normal rate limits, allowing multiple
 * instances to be used across multiple processes/machines.
 * This can be called at any time.
 */
RiotApi.prototype.setDistFactor = function(factor) {
  if (factor <= 0 || factor > 1)
    throw new Error("Factor must be greater than zero and non-greater than one.");
  if (this.config.distFactor === factor)
    return;
  this.config.distFactor = factor;
  Object.values(this.regions).forEach(r => r.updateDistFactor());
};
RiotApi.prototype._getRegion = function(region) {
  if (this.regions[region])
    return this.regions[region];
  return (this.regions[region] = new Region(this.config));
};


/** RateLimits for a region. One app limit and any number of method limits. */
function Region(config) {
  this.config = config;
  this.appLimit = new RateLimit(this.config.rateLimitTypeApplication, 1, this.config);
  this.methodLimits = {};
  this.concurrentSema = new Semaphore(this.config.maxConcurrent);
}
Region.prototype.get = function() {
  let prefix = arguments[0];
  let target = arguments[1];
  let suffix = this.config.endpoints;
  for (let path of target.split('.')) {
    suffix = suffix[path];
    if (!suffix) throw new Error(util.format('Missing path "%s" in "%s".', path, target));
  }
  let qs = {};
  let args = Array.prototype.slice.call(arguments, 2);
  if (typeof args[args.length - 1] === 'object') // If last obj is query string, pop it off.
    qs = args.pop();
  if (suffix.split('%s').length - 1 !== args.length)
    throw new Error(util.format('Wrong number of path arguments: "%j", for path "%s".', args, suffix));
  if (this.config.collapseQueryArrays) {
    qs = JSON.parse(JSON.stringify(qs)); // Clone object so we can modify without confusion.
    Object.entries(qs).forEach(([ key, val ]) => qs[key] = Array.isArray(val) ? val.join(',') : val);
  }
  suffix = util.format(suffix, ...args.map(arg => encodeURIComponent(arg)));
  let uri = prefix + suffix;

  let rateLimits = [ this.appLimit ];
  if (this.config.rateLimitTypeMethod) // Also method limit if applicable.
    rateLimits.push(this._getMethodLimit(target));
  let retries = 0;

  let fn = () => {
    let delay = RateLimit.getAllOrDelay(rateLimits);
    if (delay >= 0)
      return delayPromise(delay).then(fn);
    let reqConfig = {
      uri, qs,
      forever: true, // keep-alive.
      simple: false,
      resolveWithFullResponse: true,
      qsStringifyOptions: { arrayFormat: { indices: false }},
      timeout: this.config.timeout,
    };
    if (this.config.keyHeader)
      reqConfig.headers = { [this.config.keyHeader]: this.config.key };
    else
      qs[this.config.keyQueryParam] = this.config.key;
    return req(reqConfig).then(res => {
      rateLimits.forEach(rl => rl.onResponse(res));
      if (400 === res.statusCode)
        throw new Error('Bad request, responded with: ' + res.body + '\nurl:' + reqConfig.uri);
      if ([ 204, 404, 422 ].includes(res.statusCode))
        return null;
      if (429 === res.statusCode || 500 <= res.statusCode) {
        if (retries >= this.config.retries)
          throw new Error('Failed after ' + retries + ' retries with code ' + res.statusCode + '.');
        retries++;
        return fn();
      }
      return JSON.parse(res.body);
    }, err => {
      if (err.error && [ 'ECONNRESET', 'ESOCKETTIMEDOUT', 'ETIMEDOUT' ].includes(err.error.code)
          && retries < this.config.retries) { // Retry connection resets/timeouts.
        retries++;
        return fn();
      }
      throw new Error('Failed after ' + retries + ' retries due to: ' + err.message);
    });
  };
  return new Promise((resolve, reject) => {
    this.concurrentSema.acquire()
      .then(fn)
      .then(out => {
        this.concurrentSema.release();
        resolve(out);
      }, err => {
        this.concurrentSema.release();
        reject(err);
      });
  });
};
Region.prototype.updateDistFactor = function() {
  this.appLimit.setDistFactor(this.config.distFactor);
  Object.values(this.methodLimits).forEach(rl => rl.setDistFactor(this.config.distFactor));
};
Region.prototype._getMethodLimit = function(method) {
  if (this.methodLimits[method])
    return this.methodLimits[method];
  return (this.methodLimits[method] = new RateLimit(this.config.rateLimitTypeMethod, 1, this.config));
};


/** Rate limit. A collection of token buckets, updated when needed. */
function RateLimit(type, distFactor, config) {
  this.config = config;
  this.type = type;
  this.buckets = this.config.defaultBuckets.map(b => new TokenBucket(b.timespan, b.limit, b));
  this.retryAfter = 0;
  this.distFactor = distFactor;
}
RateLimit.prototype.retryDelay = function() {
  let now = Date.now();
  return now > this.retryAfter ? -1 : this.retryAfter - now;
};
RateLimit.prototype.onResponse = function(res) {
  if (429 === res.statusCode) {
    let type = res.headers[this.config.headerLimitType] || this.config.defaultLimitType;
    if (!type)
      throw new Error('Response missing type.');
    if (this.type.name === type.toLowerCase()) {
      let retryAfter = +res.headers[this.config.headerRetryAfter];
      if (Number.isNaN(retryAfter))
        throw new Error('Response 429 missing retry-after header.');
      this.retryAfter = Date.now() + retryAfter * 1000 + 500;
    }
  }

  const limitHeader = res.headers[this.type.headerLimit];
  const countHeader = res.headers[this.type.headerCount];
  if (this._bucketsNeedUpdate(limitHeader, countHeader))
    this.buckets = RateLimit._getBucketsFromHeaders(limitHeader, countHeader);
};
RateLimit.prototype.setDistFactor = function(factor) {
  this.distFactor = factor;
  this.buckets.forEach(b => b.setDistFactor(factor));
};
RateLimit.prototype._bucketsNeedUpdate = function(limitHeader, countHeader) {
  if (!limitHeader || !countHeader)
    return false;
  let limits = this.buckets.map(b => b.toLimitString()).join(',');
  return limitHeader !== limits;
};
RateLimit._getBucketsFromHeaders = function(limitHeader, countHeader) {
  // Limits: "20000:10,1200000:600"
  // Counts: "7:10,58:600"
  let limits = limitHeader.split(',');
  let counts = countHeader.split(',');
  if (limits.length !== counts.length)
    throw new Error('Limit and count headers do not match: ' + limitHeader + ', ' + countHeader + '.');

  return limits
    .map((limit, i) => {
      let count = counts[i];
      let [limitVal, limitSpan] = limit.split(':').map(Number);
      let [countVal, countSpan] = count.split(':').map(Number);
      limitSpan *= 1000;
      countSpan *= 1000;
      if (limitSpan != countSpan)
        throw new Error('Limit span and count span do not match: ' + limitSpan + ', ' + countSpan + '.');

      let bucket = new TokenBucket(limitSpan, limitVal, { distFactor: this.distFactor });
      bucket.getTokens(countVal);
      return bucket;
    });
};
RateLimit.getAllOrDelay = function(rateLimits) {
  let delay = rateLimits
    .map(r => r.retryDelay())
    .reduce((a, b) => Math.max(a, b), -1);
  if (delay >= 0)
    return delay; // Techincally the delay could be more but whatev.
  let allBuckets = [].concat.apply([], rateLimits.map(rl => rl.buckets));
  return TokenBucket.getAllOrDelay(allBuckets);
};

/** Token bucket. Represents a single "100:60", AKA a 100 tokens per 60 seconds pair. */
function TokenBucket(timespan, limit, { distFactor = 1, factor = 20, spread = 500 / timespan, now = Date.now } = {}) {
  this.now = now;

  this.timespan = timespan;
  // this.givenLimit is the limit given to the constructor, this.limit is the
  // functional limit, accounting for this.distFactor.
  this.givenLimit = limit;
  this.factor = factor;
  this.spread = spread;
  this.timespanIndex = Math.ceil(timespan / factor);
  this.total = 0;
  this.time = -1;
  this.buffer = new Array(this.factor + 1).fill(0);

  this.setDistFactor(distFactor);
}
TokenBucket.prototype.setDistFactor = function(distFactor) {
  this.limit = this.givenLimit * distFactor;
  // TODO: this math is ugly and basically wrong
  this.limitPerIndex = Math.floor(this.givenLimit / this.spread / this.factor) * distFactor;
  if (this.limitPerIndex * this.factor < this.limit) // TODO: hack to fix math above
    this.limitPerIndex = Math.ceil(this.limit / this.factor);
};
/** Returns delay in milliseconds or -1 if token available. */
TokenBucket.prototype.getDelay = function() {
  let index = this._update();
  if (this.total < this.limit) {
    if (this.buffer[index] >= this.limitPerIndex)
      return this._getTimeToBucket(1);
    return -1;
  }

  // check how soon into the future old buckets will be zeroed, making requests available.
  let i = 1;
  for (; i < this.buffer.length; i++) {
    if (this.buffer[(index + i) % this.buffer.length] > 0)
      break;
  }
  return this._getTimeToBucket(i);
}
TokenBucket.prototype.getTokens = function(n) {
  let index = this._update();
  this.buffer[index] += n;
  this.total += n;
  return this.total <= this.limit && this.buffer[index] <= this.limitPerIndex;
};
TokenBucket.prototype.toLimitString = function() {
  return this.givenLimit + ':' + (this.timespan / 1000);
};
TokenBucket.prototype._update = function() {
  if (this.time < 0) {
    this.time = this.now();
    return this._getIndex(this.time);
  }

  let index = this._getIndex(this.time);
  let length = this._getLength(this.time, (this.time = this.now()));

  if (length < 0)
    throw new Error('Negative length.');
  if (length == 0)
    return index;
  if (length >= this.buffer.length) {
    this.buffer.fill(0);
    this.total = 0;
    return index;
  }
  for (let i = 0; i < length; i++) {
    index++;
    index %= this.buffer.length;
    this.total -= this.buffer[index];
    this.buffer[index] = 0;
  }
  if (this._getIndex(this.time) != index)
    throw new Error('Get index time wrong:' + this._getIndex(this.time) + ', ' + index + '.');
  return index;
};
TokenBucket.prototype._getIndex = function(ts) {
  return Math.floor((ts / this.timespanIndex) % this.buffer.length);
};
TokenBucket.prototype._getLength = function(start, end) {
  return Math.floor(end / this.timespanIndex) - Math.floor(start / this.timespanIndex);
};
TokenBucket.prototype._getTimeToBucket = function(n) {
  return n * this.timespanIndex - (this.time % this.timespanIndex);
};
TokenBucket.getAllOrDelay = function(tokenBuckets) {
  let delay = tokenBuckets
    .map(b => b.getDelay())
    .reduce((a, b) => Math.max(a, b), -1);
  if (delay >= 0)
    return delay;
  tokenBuckets.forEach(b => b.getTokens(1));
  return -1;
};

function Semaphore(permits) {
  this._permits = permits;
  this._queue = [];
}
Semaphore.prototype.acquire = function() {
  return new Promise(resolve => {
    if (this._permits) {
      this._permits--;
      resolve();
    }
    else
      this._queue.push(resolve);
  });
};
Semaphore.prototype.release = function() {
  const resolve = this._queue.shift();
  (resolve ? resolve() : this._permits++);
};

module.exports = RiotApi;
if (DEBUG)
  Object.assign(module.exports, { delayPromise, TokenBucket, Semaphore });
